java.lang.VerifyError when mocking ObjectMapper

I am having problem with mocking jackson ObjectMapper

version info 
```
    $micronautVersion=1.2.7
    compileOnly "com.oracle.substratevm:svm"
    compileOnly 'org.projectlombok:lombok:1.16.20'
     annotationProcessor "org.projectlombok:lombok:1.16.20"
    annotationProcessor platform("io.micronaut:micronaut-bom:$micronautVersion")
    annotationProcessor "io.micronaut:micronaut-inject-java"
    annotationProcessor "io.micronaut:micronaut-validation"
    implementation platform("io.micronaut:micronaut-bom:$micronautVersion")
    implementation "io.micronaut:micronaut-inject"
    implementation "io.micronaut:micronaut-validation"
    implementation "io.micronaut:micronaut-runtime"
    implementation "io.micronaut.configuration:micronaut-kafka"
    runtimeOnly "ch.qos.logback:logback-classic:1.2.3"
    testAnnotationProcessor platform("io.micronaut:micronaut-bom:$micronautVersion")
    testAnnotationProcessor "io.micronaut:micronaut-inject-java"
    testImplementation "org.junit.jupiter:junit-jupiter-api"
    testImplementation "io.micronaut.test:micronaut-test-junit5"
    testRuntimeOnly "org.junit.jupiter:junit-jupiter-engine"
    testImplementation "org.mockito:mockito-junit-jupiter:2.22.0"
```

if I use 

```java 
@MockBean(ObjectMapper.class)
    ObjectMapper objectMapper(){
        return mock(ObjectMapper.class);
    }
```
then I am getting this log


`Unexpected error loading bean definition [XYZTest$ObjectMapper0Definition$InterceptedDefinition]: class XYZTest$ObjectMapper0Definition$Intercepted overrides final method readValue.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/core/type/ResolvedType;)Ljava/lang/Object;
io.micronaut.context.exceptions.BeanContextException: Unexpected error loading bean definition [XYZTest$ObjectMapper0Definition$InterceptedDefinition]: class XYZTest$ObjectMapper0Definition$Intercepted overrides final method readValue.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/core/type/ResolvedType;)Ljava/lang/Object;
	at io.micronaut.context.AbstractBeanDefinitionReference.isPresent(AbstractBeanDefinitionReference.java:93)
	at io.micronaut.context.DefaultBeanContext.indexBeanDefinitions(DefaultBeanContext.java:2451)
	at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2444)
	at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:200)
	at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:187)
	at io.micronaut.test.extensions.AbstractMicronautExtension.startApplicationContext(AbstractMicronautExtension.java:253)
	at io.micronaut.test.extensions.AbstractMicronautExtension.beforeClass(AbstractMicronautExtension.java:174)
	at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.beforeAll(MicronautJunit5Extension.java:49)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.lambda$invokeBeforeAllCallbacks$7(ClassBasedTestDescriptor.java:353)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.invokeBeforeAllCallbacks(ClassBasedTestDescriptor.java:353)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:187)
	at org.junit.jupiter.engine.descriptor.ClassBasedTestDescriptor.before(ClassBasedTestDescriptor.java:77)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:132)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at java.util.ArrayList.forEach(ArrayList.java:1257)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:38)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$5(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$7(NodeTestTask.java:125)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:135)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:123)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:122)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:80)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:32)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:51)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:220)
	at org.junit.platform.launcher.core.DefaultLauncher.lambda$execute$6(DefaultLauncher.java:188)
	at org.junit.platform.launcher.core.DefaultLauncher.withInterceptedStreams(DefaultLauncher.java:202)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:181)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:128)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:102)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:82)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:78)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:32)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:93)
	at com.sun.proxy.$Proxy2.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.stop(TestWorker.java:132)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:175)
	at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:157)
	at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
	at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
	at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.VerifyError: class XYZTest$ObjectMapper0Definition$Intercepted overrides final method readValue.(Lcom/fasterxml/jackson/core/JsonParser;Lcom/fasterxml/jackson/core/type/ResolvedType;)Ljava/lang/Object;
	at java.lang.ClassLoader.defineClass1(Native Method)
	at java.lang.ClassLoader.defineClass(ClassLoader.java:763)
	at java.security.SecureClassLoader.defineClass(SecureClassLoader.java:142)
	at java.net.URLClassLoader.defineClass(URLClassLoader.java:468)
	at java.net.URLClassLoader.access$100(URLClassLoader.java:74)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:369)
	at java.net.URLClassLoader$1.run(URLClassLoader.java:363)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.net.URLClassLoader.findClass(URLClassLoader.java:362)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:349)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at XYZTest$ObjectMapper0Definition$InterceptedDefinitionClass.getBeanType(Unknown Source)
	at io.micronaut.context.AbstractBeanDefinitionReference.isPresent(AbstractBeanDefinitionReference.java:85)
	... 67 more`

--- a/http-client/src/test/groovy/io/micronaut/http/client/DataStreamSpec.groovy
+++ b/http-client/src/test/groovy/io/micronaut/http/client/DataStreamSpec.groovy
@@ -42,6 +42,7 @@ import org.reactivestreams.Subscriber
 import org.reactivestreams.Subscription
 import spock.lang.AutoCleanup
 import spock.lang.Issue
+import spock.lang.Retry
 import spock.lang.Shared
 import spock.lang.Specification
 import spock.util.concurrent.PollingConditions
@@ -131,6 +132,7 @@ class DataStreamSpec extends Specification {
         new String(arrays[1]) == 'The Shining'
     }
 
+    @Retry
     void "test that stream response is free of race conditions"() {
         when:
         List<byte[]> arrays = client.exchangeStream(HttpRequest.GET(

--- a/inject-java/src/main/java/io/micronaut/annotation/processing/SuperclassAwareTypeVisitor.java
+++ b/inject-java/src/main/java/io/micronaut/annotation/processing/SuperclassAwareTypeVisitor.java
@@ -20,6 +20,7 @@ import io.micronaut.inject.processing.JavaModelUtils;
 
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.*;
 import javax.lang.model.util.AbstractTypeVisitor8;
@@ -65,22 +66,7 @@ public abstract class SuperclassAwareTypeVisitor<R, P> extends AbstractTypeVisit
                 boolean isAcceptable = isAcceptable(enclosedElement);
                 if (isAcceptable) {
                     if (enclosedElement instanceof ExecutableElement) {
-                        ExecutableElement ee = (ExecutableElement) enclosedElement;
-                        String qualifiedName = ee.getSimpleName().toString();
-                        qualifiedName += "(" + ee.getParameters().stream().map(variableElement -> types.erasure(variableElement.asType()).toString()).collect(Collectors.joining(",")) + ")";
-
-
-                        TypeMirror returnTypeMirror = ee.getReturnType();
-                        String returnType = types.erasure(returnTypeMirror).toString();
-                        if (returnTypeMirror.getKind() == TypeKind.TYPEVAR) {
-                            Map<String, Object> generics = genericUtils.buildGenericTypeArgumentInfo(type)
-                                    .get(typeElement.getQualifiedName().toString());
-                            if (generics != null && generics.containsKey(returnTypeMirror.toString())) {
-                                returnType = generics.get(returnTypeMirror.toString()).toString();
-                            }
-                        }
-
-                        qualifiedName = returnType + "." + qualifiedName;
+                        String qualifiedName = buildQualifiedName(type, typeElement, (ExecutableElement) enclosedElement);
 
                         // if the method has already been processed then it is overridden so ignore
                         if (!processed.contains(qualifiedName)) {
@@ -94,6 +80,15 @@ public abstract class SuperclassAwareTypeVisitor<R, P> extends AbstractTypeVisit
                             processed.add(qualifiedName);
                             accept(type, enclosedElement, p);
                         }
+                    }
+                } else if (enclosedElement instanceof ExecutableElement) {
+                    ExecutableElement ee = (ExecutableElement) enclosedElement;
+                    Set<Modifier> modifiers = ee.getModifiers();
+                    if (!modifiers.contains(Modifier.PRIVATE) && !modifiers.contains(Modifier.STATIC)) {
+                        String qualifiedName = buildQualifiedName(type, typeElement, ee);
+                        // add to processed so that if a super method is visited that this method
+                        // overrides it is not processed again
+                        processed.add(qualifiedName);
                     }
                 }
             }
@@ -177,5 +172,25 @@ public abstract class SuperclassAwareTypeVisitor<R, P> extends AbstractTypeVisit
     @Override
     public R visitUnion(UnionType t, P p) {
         return null;
+    }
+
+    private String buildQualifiedName(DeclaredType type, TypeElement typeElement, ExecutableElement enclosedElement) {
+        ExecutableElement ee = enclosedElement;
+        String qualifiedName = ee.getSimpleName().toString();
+        qualifiedName += "(" + ee.getParameters().stream().map(variableElement -> types.erasure(variableElement.asType()).toString()).collect(Collectors.joining(",")) + ")";
+
+        TypeMirror returnTypeMirror = ee.getReturnType();
+        String returnType = types.erasure(returnTypeMirror).toString();
+
+        if (returnTypeMirror.getKind() == TypeKind.TYPEVAR) {
+            Map<String, Object> generics = genericUtils.buildGenericTypeArgumentInfo(type)
+                    .get(typeElement.getQualifiedName().toString());
+            if (generics != null && generics.containsKey(returnTypeMirror.toString())) {
+                returnType = generics.get(returnTypeMirror.toString()).toString();
+            }
+        }
+
+        qualifiedName = returnType + "." + qualifiedName;
+        return qualifiedName;
     }
 }

--- a/inject-java/src/test/groovy/io/micronaut/aop/compile/FinalModifierSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/aop/compile/FinalModifierSpec.groovy
@@ -15,11 +15,43 @@
  */
 package io.micronaut.aop.compile
 
+import com.fasterxml.jackson.databind.ObjectMapper
+import io.micronaut.aop.Intercepted
 import io.micronaut.aop.simple.Mutating
 import io.micronaut.inject.AbstractTypeElementSpec
+import io.micronaut.inject.BeanDefinition
+import io.micronaut.inject.qualifiers.Qualifiers
 import spock.lang.Issue
 
 class FinalModifierSpec extends AbstractTypeElementSpec {
+
+    @Issue('https://github.com/micronaut-projects/micronaut-core/issues/2530')
+    void 'test final modifier on external class produced by factory'() {
+        when:
+        def context = buildContext('test.MyBeanFactory', '''
+package test;
+
+import io.micronaut.aop.simple.*;
+import io.micronaut.context.annotation.*;
+import com.fasterxml.jackson.databind.ObjectMapper;
+@Factory
+class MyBeanFactory {
+    @Mutating("someVal")
+    @javax.inject.Singleton
+    @javax.inject.Named("myMapper")
+    ObjectMapper myMapper() {
+        return new ObjectMapper();
+    }
+
+}
+
+''')
+        then:
+        context.getBean(ObjectMapper, Qualifiers.byName("myMapper")) instanceof Intercepted
+
+        cleanup:
+        context.close()
+    }
 
     @Issue('https://github.com/micronaut-projects/micronaut-core/issues/2479')
     void "test final modifier on inherited public method"() {

